src/fstext/determinize-star-inl.h

```diff
@@ -468,99 +469,232 @@ template<class Arc> class DeterminizerStar {
+    EpsilonClosureInfo() {}
+
+    EpsilonClosureInfo(Element e, Weight w, bool i, int d = 0) :
+      element(e), weight_to_process(w), in_queue(i), depth(d) {}
+
+    // the weight in the Element struct is the total current weight
+    // that has been processed already
+    Element element;
+
+    // this stores the weight that we haven't processed (propagated)
+    Weight weight_to_process;
+
+    // whether "this" struct is in the queue
+    // we store the info here so that we don't have to look it up every time
+    bool in_queue;
+
+    // depth is the max number of epsilons (so far) from the original element set
+    // e.g. if the original set is (s1, s2)
+    // s1 -eps-> s4
+    // s2 -eps-> s3 -eps-> s4
+    // then the depth for s4 would eventually be 2
+
+    // to avoid duplicate of computation, we use a priority_queue for storing
+    // the state_id's to process, and always pick the one with the smallest
+    // depth so far first
+    //
+    // this would be similar to topo-sort the states before processing;
+    // (not quite identical since we don't compute the definitive depth first,
+    // but gradually update the depth while running
+    // it seems to speed up the process very significantly already
+    int depth;
+  };
+
+  // this is the struct that is put into the priority_queue. 
+  // The comparison operator makes sure that the queue.top() returns
+  // the one with the least depth
+  struct IdAndDepth {
+    IdAndDepth (InputStateId i, int d):
+      id(i), depth(d) {}
+    InputStateId id;
+    int depth;
+    bool operator <(const IdAndDepth& other) const {
+      return this->depth > other.depth;  // we want the lowest depth on top
+    }
+  };
+  
+  // queue for EpsilonClosure() function
+  // put it here so the memeory allocation for better effiency
+  // otherwise it gets created and destroyed constantly, also
+  // possibly a lot of resize() call the for underlying vector
+  priority_queue<IdAndDepth> queue;
+
+  // Sub-routine that we call in EpsilonClosure()
+  // To avoid the function getting too long, we make it a separate function
+  // It takes the current "unprocessed_weight" and propagate it to the 
+  // states accessible from elem.state by an epsilon arc
+  // and add the results to cur_subset.
+  void ExpandOneElement(const Element &elem,
+                        bool sorted,
+                        Weight unprocessed_weight,
+                        std::map<InputStateId, EpsilonClosureInfo> *cur_subset,
+                        int this_depth) {
+    // now we are going to propagate the "unprocessed_weight"
+    for (ArcIterator<Fst<Arc> > aiter(*ifst_, elem.state);
+         !aiter.Done(); aiter.Next()) {
+      const Arc &arc = aiter.Value();
+      if (sorted && arc.ilabel > 0) {
+        break;
+        // Break from the loop: due to sorting there will be no
+        // more transitions with epsilons as input labels.
+      }
+      if (arc.ilabel == 0) {  // Epsilon transition.
```

Comment:
You could make the indentation level more reasonable by changing this to
if (arc.ilabel != 0)
  continue;  // We are only processing epsilons here.